hledger/Hledger/Cli/Commands/Run.hs

```diff
@@ -108,5 +115,46 @@ runREPL findBuiltinCommand j = do
       Just "quit" -> return ()
       Just "exit" -> return ()
       Just input -> do
-        liftIO $ runCommand findBuiltinCommand j $ parseCommand input
+        liftIO $ runCommand findBuiltinCommand $ parseCommand input
         loop
+
+{-# NOINLINE journalCache #-}
+journalCache :: MVar (Map.Map [String] Journal)
+journalCache = unsafePerformIO $ newMVar Map.empty
+
+-- | Key used to cache the journal given in the arguments to 'run'.
+defaultJournalKey :: [String]
+defaultJournalKey = ["journal specified in args of run"]
+
+hasStdin :: [String] -> Bool
+hasStdin fps = "-" `elem` fps
+
+addJournalToCache :: Journal -> [String] -> IO ()
+addJournalToCache j journalpaths = modifyMVar_ journalCache $ \cache ->
+  if hasStdin $ dbg1 "addJournalToCache" journalpaths
+  then return cache
+  else return $ Map.insert journalpaths j cache
+
+withJournalCached :: CliOpts -> (Journal -> IO ()) -> IO ()
+withJournalCached cliopts f = do
+  journalpaths <- journalFilePathFromOptsNoDefault cliopts
```

Comment:
I am a little confused, doesn't this return a single file path ? Wouldn't "journalpath" or "p" be clearer ? Or following my usual convention, "mp" (maybe p).

Reply:
This returns a list, if there are several `-f` options. I've called it `mbjournalpaths` in later commit